# Homework 2: Experiment tracing with MLflow

Question [here](https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/02-experiment-tracking/homework.md)

## Question 1. Install MLflow

Run the command mlflow --version and check the output.

What's the version that you have?

In [2]:
!mlflow --version

mlflow, version 1.26.1


### Answer 1.
Version 1.26.1

## Question 2. Download and preprocess the data

Use the script preprocess_data.py located in the folder homework to preprocess the data.

The script will

* load the data from the folder <TAXI_DATA_FOLDER> (the folder where you have downloaded the data),
* fit a DictVectorizer on the training set (January 2021 data),
* save the preprocessed datasets and the model to disk

In [3]:
!python preprocess_data.py --raw_data_path ../data --dest_path ./output

In [6]:
%ls output

dv.pkl  test.pkl  train.pkl  valid.pkl


### Answer 2. 
4 files were saved

## Question3. Train a model with autolog
Modify the given script to enable autologging with MLflow, execute the script and then launch the MLflow UI to check that the experiment run was properly tracked.

In [12]:
!python train.py

2022/05/29 23:41:01 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/home/jaimeh94/anaconda3/envs/mlops-zoomcamp/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."


### Answer 3.
17 parameters

## Question 4. Launch the tracking server locally

Your task is to

launch the tracking server on your local machine
select a SQLite db for the backend store and a folder called artifacts for the artifacts store
You should keep the tracking server running to work on the next two exercises that use the server.

In addition to backend-store-uri, what else do you need to pass to properly configure the server?


In [13]:
!mlflow ui --help

Usage: mlflow ui [OPTIONS]

  Launch the MLflow tracking UI for local viewing of run results. To launch a
  production server, use the "mlflow server" command instead.

  The UI will be visible at http://localhost:5000 by default, and only accepts
  connections from the local machine. To let the UI server accept connections
  from other machines, you will need to pass ``--host 0.0.0.0`` to listen on
  all network interfaces (or a specific interface address).

Options:
  --backend-store-uri PATH     URI to which to persist experiment and run
                               data. Acceptable URIs are SQLAlchemy-compatible
                               database connection strings (e.g.
                               'sqlite:///path/to/file.db') or local
                               filesystem URIs (e.g.
                               'file:///absolute/path/to/directory'). By
                               default, data will be logged to ./mlruns
  --default-artifact-root URI  Directory i

### Answer 4.
--default-artifact-root

## Question 5. Tune the hyperparameters of the model
Your task is to modify the script hpo.py and make sure that the validation RMSE is logged to MLflow for each run of the hyperparameter optimization (you will need to add a few lines of code to the objective function) and run the script without passing any parameters.

What's the best validation RMSE that you got?

In [16]:
!python hpo.py

100%|█████████| 50/50 [04:19<00:00,  5.19s/trial, best loss: 6.6284257482044735]


### Answer 5.
6.628

## Question 6. Promote the best model to the model registry

Your task is to update the script register_model.py so that it selects the model with the lowest RMSE on the test set and registers it to the model registry.

What is the test RMSE of the best model?

In [ ]:
!python register_model.py

### Answer 6.
6.55